src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/x64/sysv/SysVx64Linker.java

```diff
@@ -51,20 +54,73 @@
 
     static final long ADDRESS_SIZE = 64; // bits
 
+    private static final MethodHandle MH_unboxVaList;
+    private static final MethodHandle MH_boxVaList;
+
+    static {
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            MH_unboxVaList = lookup.findStatic(SysVx64Linker.class, "unboxVaList",
+                MethodType.methodType(MemoryAddress.class, CSupport.VaList.class));
+            MH_boxVaList = lookup.findStatic(SysVx64Linker.class, "boxVaList",
+                MethodType.methodType(VaList.class, MemoryAddress.class));
+        } catch (ReflectiveOperationException e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
+
     public static SysVx64Linker getInstance() {
         if (instance == null) {
             instance = new SysVx64Linker();
         }
         return instance;
     }
 
+    public static VaList newVaList(Consumer<VaList.Builder> actions) {
+        SysVVaList.Builder builder = SysVVaList.builder();
+        actions.accept(builder);
+        return builder.build();
+    }
+
+    private static MethodType convertVaListCarriers(MethodType mt) {
+        Class<?>[] params = new Class<?>[mt.parameterCount()];
+        for (int i = 0; i < params.length; i++) {
+            Class<?> pType = mt.parameterType(i);
+            params[i] = ((pType == CSupport.VaList.class) ? SysVVaList.CARRIER : pType);
+        }
+        return MethodType.methodType(mt.returnType(), params);
+    }
+
+    private static MethodHandle unxboxVaLists(MethodType type, MethodHandle handle) {
+        for (int i = 0; i < type.parameterCount(); i++) {
+            if (type.parameterType(i) == VaList.class) {
+               handle = MethodHandles.filterArguments(handle, i, MH_unboxVaList);
+            }
+        }
+        return handle;
+    }
+
     @Override
     public MethodHandle downcallHandle(MemoryAddress symbol, MethodType type, FunctionDescriptor function) {
-        return CallArranger.arrangeDowncall(symbol, type, function);
+        MethodType llMt = convertVaListCarriers(type);
+        MethodHandle handle = CallArranger.arrangeDowncall(symbol, llMt, function);
+        handle = unxboxVaLists(type, handle);
+        return handle;
+    }
+
+    private static MethodHandle boxVaLists(MethodHandle handle) {
+        MethodType type = handle.type();
+        for (int i = 0; i < type.parameterCount(); i++) {
+            if (type.parameterType(i) == VaList.class) {
+               handle = MethodHandles.filterArguments(handle, i, MH_boxVaList);
+            }
```

Comment:
These looks a bit odd - in the sense that we have to hide valists from call arranger, but then add adaptation back in to make clients happy. I guess it seems as if this could be less error prone if CallArranger would also be in charge of dealing with valists.